File: week4/1_phone_book/phone_book.py
# python3


# Use the direct addressing scheme to solve this problem
class Query:
    def __init__(self, query):
        # add, find or delete query?
        self.type = query[0]
        self.number = int(query[1])
        if self.type == 'add':
            self.name = query[2]


def read_queries():
    n = int(input())
    return [Query(input().split()) for _ in range(n)]


def write_responses(result):
    print('\n'.join(result))


def process_queries(queries):
    result = []
    # list of contacts
    contacts = dict()
    for this in queries:
        if this.type == 'add':
            if this.number in contacts:
                contacts[this.number] = this.name
            else:
                contacts[this.number] = this.name
        elif this.type == 'del':
            if this.number in contacts:
                del contacts[this.number]
        else:
            if this.number in contacts:
                result.append(contacts[this.number])
            else:
                result.append('not found')
    return result


if __name__ == '__main__':
    write_responses(process_queries(read_queries()))

File: week4/1_phone_book/phone_book.cpp
#include <iostream>
#include <vector>
#include <string>
#include <map>

using namespace std;

struct Query {
    string type, name;
    int number;
};

vector<Query> read_queries() {
    int n;
    cin >> n;
    vector<Query> queries(n);
    for (int i = 0; i < n; ++i) {
        cin >> queries[i].type;
        if (queries[i].type == "add")
            cin >> queries[i].number >> queries[i].name;
        else
            cin >> queries[i].number;
    }
    return queries;
}

void write_responses(const vector<string>& result) {
    for (size_t i = 0; i < result.size(); ++i)
        std::cout << result[i] << "\n";
}

vector<string> process_queries(const vector<Query>& queries) {
    vector<string> result;
    // list of contacts
    map<int, string> contacts;
    map<int, string>::iterator found;
    for (size_t i = 0; i < queries.size(); ++i) {
        if (queries[i].type == "add") {
            bool was_founded = false;
            // if current in contacts, rewrite contact's name
            found = contacts.find(queries[i].number);
            if (found != contacts.end()) {
                found->second = queries[i].name;
                was_founded = true;
            }
            // otherwise, just add it
            if (!was_founded)
                contacts[queries[i].number] = queries[i].name;
        } else if (queries[i].type == "del") {
            found = contacts.find(queries[i].number);
            if (found != contacts.end()) {
                contacts.erase(found);
            }
        } else {
            string response = "not found";
            found = contacts.find(queries[i].number);
            if (found != contacts.end()) {
                response = found->second;
            }
            result.push_back(response);
        }
    }
    return result;
}

int main() {
    write_responses(process_queries(read_queries()));
    return 0;
}




